hiero-observability/hiero-metrics-core/src/main/java/org/hiero/metrics/api/core/MetricRegistry.java

```diff
@@ -0,0 +1,204 @@
+// SPDX-License-Identifier: Apache-2.0
+package org.hiero.metrics.api.core;
+
+import com.swirlds.base.ArgumentUtils;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.hiero.metrics.api.utils.MetricUtils;
+import org.hiero.metrics.internal.core.MetricRegistryImpl;
+
+/**
+ * A thread-safe registry for {@link Metric} instances.
+ * <p>
+ * Registry can be created via {@link #builder(String)} using builder pattern.
+ */
+public sealed interface MetricRegistry permits org.hiero.metrics.internal.export.SnapshotableMetricsRegistry {
+
+    /**
+     * Returns the name of the metric registry.
+     *
+     * @return the name of the registry, never {@code null}
+     */
+    @NonNull
+    String name();
+
+    /**
+     * Returns an unmodifiable list of global labels that are applied to all metrics in this registry.
+     *
+     * @return the global labels, may be empty but never {@code null}
+     */
+    @NonNull
+    List<Label> globalLabels();
+
+    /**
+     * Returns an unmodifiable collection of all registered metrics.
+     * Order of the metrics is not guaranteed and can change over time as new metrics are registered.
+     *
+     * @return the registered metrics, may be empty but never {@code null}
+     */
+    @NonNull
+    Collection<Metric> metrics();
+
+    /**
+     * Registers a metric using the given builder.
+     * <p>
+     * This method is <b>not idempotent</b> and throws an exception if metrics with same name already registered.
```

Comment:
Would it be better/easier if this method was just a no-op, if the metric has been already registered rather than throwing exceptions?